[jamesebond] iP

jamesebond opened the pull request:
# JamesBond

> The name's Bond, James Bond. :shipit:

## Description
This individual Project (iP) JamesBond frees your mind of having to remember things you need to do. This particular Pull request (PR) has, 

- Added features like adding and deleting tasks
- Added search function
- Stores task history

**_WOW_!**

Includes: 
- [x] 🐛 fixes
- [ ] Improved performance

All you need to do for more additional information is

1.  Click [here](https://github.com/jamesebond/ip)
2. Enjoy the new features!!!!

**Code Changes in `main` method:**
```
package duke;

import exceptions.JamesBondException;
import parser.Parser;
import storage.Storage;
import tasks.TaskList;
import ui.Ui;
import commands.*;

/**
 * The `Duke` class is the main class for the Duke task management application.
 * It handles the initialization of various components and the execution of the application.
 */
public class Duke {

    private Storage storage;
    private TaskList tasks;
    private Ui ui;

    /**
     * Constructs a `Duke` object with the specified file path for data storage.
     *
     * @param filePath The file path where task data is stored.
     */
    public Duke(String filePath) {
        ui = new Ui();
        storage = new Storage(filePath);
        try {
            TaskList tasks = new TaskList();
            tasks = storage.loadTasksFromFile();
            this.tasks = tasks;
        } catch (JamesBondException e) {
            ui.showLoadingError();
            tasks = new TaskList();
        }
    }

    /**
     * Runs the Duke application.
     * It displays a welcome message and continuously reads user input, parsing and executing commands
     * until there is no more input.
     */
    public void run() {
        ui.showWelcomeMessage();
        while (ui.isThereNext()) {
            String input = ui.readLine();
            Command c = Parser.parseCommand(input);
            c.runCommand(tasks, ui, storage);
        }
    }

    /**
     * The main method to start the Duke application.
     * It creates an instance of `Duke` and runs the application.
     *
     * @param args Command-line arguments (not used in this application).
     */
    public static void main(String[] args) {
        new Duke("/Users/jamesbond/ip/data/jamesbond.txt").run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

jamesebond committed:
Level 0. Rename, Greet, Exit

jamesebond committed:
Level 1. Echo

jamesebond committed:
Level 2: Add, List

jamesebond committed:
Level 3. Mark as Done

jamesebond committed:
Level 4. ToDos, Events, Deadlines

jamesebond committed:
Level 5. Handle Errors

jamesebond committed:
level 6 delte

jamesebond committed:
Add branch-Level-7

jamesebond committed:
Add Level 8 Dates and times

jamesebond committed:
Add UI

jamesebond committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add branch-Level-7

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/JamesBond.java
#	src/main/java/TaskList.java

jamesebond committed:
Delete files

jamesebond committed:
Merge branch 'branch-Level-8' into branch-Level-7

* branch-Level-8:
  Add UI
  Add Level 8 Dates and times

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/JamesBond.java
#	src/main/java/TaskList.java

jamesebond committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add UI

# Conflicts:
#	src/main/java/JamesBond.java

jamesebond committed:
Merge branch 'branch-Level-7'

* branch-Level-7:

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/JamesBond.java
#	src/main/java/TaskList.java

jamesebond committed:
Delete files

jamesebond committed:
Add A-MoreOOP

jamesebond committed:
Add jamesbond package

jamesebond committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Bump gradle and lib version
  Add Gradle support

jamesebond committed:
Add JUnit tests

jamesebond committed:
Add gitignore

jamesebond committed:
Add JavaDoc

jamesebond committed:
Add A-CodingStandard

jamesebond committed:
Add Level-9 Find

jamesebond committed:
Resolve first conflict

jamesebond committed:
Merge tag 'Level-9'

* tag 'Level-9':
  Add Level-9 Find

# Conflicts:
#	src/main/java/parser/Parser.java
#	src/main/java/tasks/Task.java
#	src/main/java/ui/Ui.java

jamesebond committed:
Merge tag 'A-CodingStandard'

* tag 'A-CodingStandard':
  Add A-CodingStandard

# Conflicts:
#	src/main/java/commands/IncorrectCommand.java
#	src/main/java/parser/Parser.java

jinyang628 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,267 @@
+            case "listout":
+                c = checkListout(split);
+                break;
+            default:
```
Love the switch statements as they improve clarity. However, there is no indentation for case clauses.

jinyang628 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,267 @@
+     * @return A Command object representing the processed command.
+     */
+    private static Command checkFind(String[] split) {
+        if(split.length == 1 || split[1].isEmpty()){
```
Perhaps leave a space after the "if" and before the "{" to adhere to our module's coding standards!

jinyang628 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,267 @@
+     * @param split The user input string split into tokens.
+     * @return A "MarkCommand" if the input is a valid mark command, otherwise an "IncorrectCommand."
+     */
+    private static Command checkMark(String[] split) {
```
Love the fact that you named you functions/methods using a verb!

jinyang628 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @param file The file path where task data is stored.
+     */
+    public Storage(String file) {
```
I believe the indentation is slightly off here

jinyang628 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,267 @@
+package parser;
+
+import commands.*;
```
Try not to use wildcard imports and import the files explicitly.

laurenlim2112 replied:
Should imported classes be listed explicitly?

jinyang628 reviewed (commented):
Love the way you use OOP throughout your program. Some minor fixes and it will be perfect!

laurenlim2112 commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+ * such as adding, deleting, retrieving, and finding tasks.
+ */
+public class TaskList {
+    ArrayList<Task> toDos;
```
I like how you used the plural form in this name!

laurenlim2112 commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+     *
+     * @return The number of tasks in the task list.
+     */
+    public int len() {
```
Perhaps this method name should be a verb?

laurenlim2112 commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,169 @@
+     *
+     * @param task The deadline task that was added.
+     */
+    public void DeadlineMessage(Task task) {
```
Should this method name be written in camelCase?

laurenlim2112 reviewed (commented):
Great job on the iP so far. However, I did note a few minor coding standard violations. Overall though, your code was easy to read, so well done!

jamesebond committed:
Made minor edits

jamesebond committed:
Add checkStyle

jamesebond committed:
Minor checkstyle edits

jamesebond committed:
Add Level-10

jamesebond committed:
Add assertions for input validation and task number range

In this commit, I have added assertions to the DeleteCommand class to improve input validation and ensure that task numbers are within a valid range:

1. In the constructor, I added an assertion to check that the task number provided is a positive integer. This helps prevent the creation of DeleteCommand objects with invalid task numbers.

2. In the runCommand method, I added assertions to validate the task number and ensure that it falls within the valid range of task indices. This check helps catch out-of-range task numbers before they lead to errors.

These assertions are designed to enhance code robustness and assist in debugging potential issues during development and testing.

jamesebond committed:
Refactor code for improved readability and maintainability

jamesebond committed:
Merge pull request #2 from jamesebond/A-CodeQuality

Refactor code for improved readability and maintainability

jamesebond committed:
Merge tag 'A-Assertions'

* tag 'A-Assertions':
  Add assertions for input validation and task number range

jamesebond committed:
Add UndoCommand for reverting the most recent action

In this commit, I've introduced the `UndoCommand` class, which provides the functionality to undo the most recent action in the task management application. When executed, this command removes the latest task addition from the task list, allowing users to revert their last change.

Changes in this commit:
- Added the `UndoCommand` class.
- Implemented the `runCommand` method to handle the undo operation.
- Included `javadoc` comments for class and method documentation.

This command enhances the user experience by offering a convenient way to correct mistakes or revert unintended changes.

jamesebond committed:
Add minor bug fixes

jamesebond committed:
Add README

jamesebond committed:
Edit README

jamesebond committed:
Fix some bugs

jamesebond committed:
Fix bugs

jamesebond committed:
Edit storage

jamesebond committed:
Minor bug fixes
